scipy/sparse/coo.py

```diff
@@ -380,16 +380,22 @@ def todok(self):
         return dok
 
     def diagonal(self):
-        # Could be rewritten without the python loop.
-        # Data entries at the same (row, col) are summed.
-        n = min(self.shape)
-        ndata = self.data.shape[0]
-        d = np.zeros(n, dtype=self.dtype)
-        for i in xrange(ndata):
-            r = self.row[i]
-            if r == self.col[i]:
-                d[r] += self.data[i]
-        return d
+        diag_ind = np.where(self.row == self.col)[0]
```

Comment:
Why not leave this as a logical index? As in:

    diag_mask = self.row == self.col